### ISSUE TITLE
Implement HLS master playlist update probing

### ISSUE THREAD

**Issue by sandersaares** [google/shaka-player#1370]
I have multi-period DASH live streams where the structure of the presentation can change substantially (entirely) between periods. Different languages, different number of adaptation sets, etc.

I am also simultaneously outputting this stream as HLS, for iOS compatibility. However, HLS does not have meaningful multiperiod support (at least when it comes to altering the presentation structure). Therefore, what I do at period changes is to abandon the existing HLS presentation and replace the master playlist with one that contains references to media playlists of the next period.

As Safari never reloads the master playlist, currently the client app has to manually monitor playback events and probe for master playlist updates so it can see this "manual period change" happening and know when to queue the playback of the new period.

It would be great if Shaka player could include a feature to probe for HLS master playlist updates in a similar manner than is done for MPD update probing with dynamic manifests. Driven by a configuration field similar in spirit to `MPD@minimumUpdatePeriod`, perhaps? Or perhaps even more useful would be server driven via `#EXT-X-MINIMUM-UPDATE-PERIOD=123` or similar.

This would simplify the duties of the host app and enable something resembling sensible multiperiod with HLS. My needs for HLS are exclusive to iOS support and so also depend on #997.

I will edit this feature request and provide a suitable test content URL in the days to come, assuming this does not sound completely out of scope for Shaka.

**Comment by ismena** [comment]
Hi!
Just out of curiosity: what is the use case for such a drastic change in the presentation?

As for updating the master playlist: I think this is doable, although it won't be a trivial change to the code.
We could include a configuration that will regulate whether or not we update the master playlist.
Then we will need to:
- decide whether we can assume we can stop updating the 'old' variants once a master playlist has changed
- make appropriate changes to the hls parser
- have the manifest parser change a list of variants and text tracks and communicate the change to the player so UI can be updated

@joeyparrish is there anything I'm missing?
Is this something you'd like to put on our road map?

**Comment by sandersaares** [comment]
> what is the use case for such a drastic change in the presentation?

This is a live stream scenario where the content has different parameters in different time periods.

For example, changing the set of audio/text languages available (consider a TV channel showing different movies - not every language might be available for each one). The languages available for ads might also differ from the langauges available for the main content.

Likewise, changing the set of quality levels offered - not all content might be available in all quality levels, or the bitrates might just be optimized differently per movie.

**Comment by joeyparrish** [comment]
It sounds like what we're doing (not updating the master playlist, only the media playlists) is in line with Safari's native HLS implementation.  The spec also doesn't mention anything about updating the master playlist, though it is explicit about updating media playlists.

For example, [RFC 8216](https://tools.ietf.org/html/rfc8216), section 6.3 ("Client Responsibilities") talks about loading the master playlist only once, under 6.3.1 ("General Client Responsibilities").  Meanwhile, reloading the media playlist gets mentioned not only under 6.3.2 ("Loading the Media Playlist File"), but also under it's own section 6.3.4 ("Reloading the Media Playlist File").

Since it seems that we are both in compliance with the spec and matching Safari's own canonical HLS client in this regard, I am going to close this issue and tag it "will not implement".

We apologize for any inconvenience.
